articles/actions.md

```diff
@@ -0,0 +1,363 @@
+
+1. Goal
+
+   This describes what the action should achieve and how it should do it.
+   It is sent to the action server when it is requested to execute an action.
+
+1. Result
+
+   This describes the outcome of an action.
+   It is sent from the server to the client when the action execution ends, whether successfully or not.
+
+1. Feedback
+
+   This describes the progress towards completing an action.
+   It is sent to the client of the action from the action server between commencing action execution and prior to the action completing.
+   This data is used by the client to understand the progress of executing the action.
+
+Any of these sections may be empty.
+Between each of the three sections is a line containing three hyphens, `---`.
+Action specifications are stored in a file ending in `.action`.
+There is one action specification per `.action` file.
+
+### Example
+
+```
+# Define a goal of washing all dishes
+bool heavy_duty  # Spend extra time cleaning
+---
+# Define the result that will be published after the action execution ends.
+uint32 total_dishes_cleaned
+---
+# Define a feedback message that will be published during action execution.
+float32 percent_complete
+uint32 number_dishes_cleaned
+```
+
+## Introspection tools
+
+Actions, like topics and services, are introspectable from the command line.
+
+The command line tool, `ros2 action`, will be able to:
+
+- list action names provided by action servers
+- list action servers and action clients
+- display active goals on an action server
+- display the arguments for an action goal
+- display the type of an action's goal, feedback, and result
+- find actions by action type
+- echo feedback, status, and result for an action goal
+- call an action, display feedback as it is received, display the result when received, and cancel the action (when the tool is terminated prematurely).
+
+Each action will be listed and treated as a single unit by this tool.
+
+## Goal States
+
+The action server maintains a state machine for each goal it accepts from a client.
+Rejected goals are not part of the state machine.
+
+![Action Goal State Machine](../img/actions/goal_state_machine.png)
+
+There are three active states:
+
+- **ACCEPTED** - The goal has been accepted and is awaiting execution.
+- **EXECUTING** - The goal is currently being executed by the action server.
+- **CANCELING** - The client has requested that the goal be canceled and the action server has accepted the cancel request.
+This state is useful for any "clean up" that the action server may have to do.
+
+And three terminal states:
+
+- **SUCCEEDED** - The goal was achieved successfully by the action server.
+- **ABORTED** - The goal was terminated by the action server without an external request.
+- **CANCELED** - The goal was canceled after an external request from an action client.
+
+State transitions triggered by the action server:
+
+- **execute** - Start execution of an accepted goal.
```

Comment:
This seems like a feature that could be built on top of actions (e.g. "real-time actions") that would involve an additional service call for a client-triggered transition. It doesn't seem necessary for the general case.

Reply:
> Are you thinking of a different action server implementation that provides a service call to initiate the transition rather than using its own timing?

Yes, exactly. The different action server (and action client) implementation could wrap the base implementation for actions so they can leverage the existing logic. But, the action server could not make the call to transition to EXECUTING until a client request to do so is received.